lib/platform/linux-uart.c

```diff
@@ -0,0 +1,437 @@
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+const char *dbginfo_off = "pscdbg 0 all\r";
+const char *echo_off = "echo 0\r";
+const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(50)
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	int idx;
+	int rty_cnt = 0;
+	char rtn[1024];
+	int cnt = 0;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = write(udev->fd, str, strlen(str));
+	if (ret != strlen(str))
+		return -1;
+
+cli_rtn_retry:
+	if (rty_cnt > RETRY_NUM)
+		return -1;
+
+	ret = read(udev->fd, rtn + cnt, sizeof(rtn));
+	if (ret < 0)
+		return ret;
+
+	cnt += ret;
+	rtn[cnt] = '\0';
+
+	if (sscanf(rtn, "%*[^:]:%x>", &idx) != 1){
+		usleep(5000);
+		rty_cnt++;
+		goto cli_rtn_retry;
+	}
+
+	return 0;
+}
+
+#ifdef __CHECKER__
+#define __force __attribute__((force))
+#else
+#define __force
+#endif
+
+static void uart_close(struct switchtec_dev *dev)
+{
+	struct switchtec_uart *udev =  to_switchtec_uart(dev);
+	cli_control(dev, echo_on);
+
+	if (dev->gas_map)
+		munmap((void __force *)dev->gas_map, dev->gas_map_size);
+
+	flock(udev->fd, LOCK_UN);
+	close(udev->fd);
+	free(udev);
+}
+
+static int map_gas(struct switchtec_dev *dev)
+{
+	void *addr;
+	dev->gas_map_size = 4 << 20;
+
+	addr = mmap(NULL, dev->gas_map_size, PROT_NONE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (addr == MAP_FAILED)
+		return -1;
+
+	dev->gas_map = (gasptr_t __force)addr;
+
+	return 0;
+}
+
+#undef __force
+
+static gasptr_t uart_gas_map(struct switchtec_dev *dev, int writeable, size_t *map_size)
+{
+	if (map_size)
+		*map_size = dev->gas_map_size;
+
+	return dev->gas_map;
+}
+
+
+static void uart_memcpy_from_gas(struct switchtec_dev *dev, void *dest, const void __gas *src, size_t n)
+{
+	int ret, raddr, rnum, rcrc, idx;
+	int rtn_rty_cnt = 0;
+	int crc_rty_cnt = 0;
+	int cnt = 0;
+	char *tok;
+	const char *delims = "]";
+	char gas_rd[1024];
+	char gas_rd_rtn[102400];
+	char *b = gas_rd;
+	uint8_t crc_cal;
+	uint8_t crc_buf[2*n];
+	uint8_t *crc_ptr = crc_buf;
+	uint32_t crc_cnt;
+	uint32_t crc_addr;
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+	uint32_t addr = (uint32_t)(src - (void __gas *)dev->gas_map);
+
+	b += snprintf(b, sizeof(gas_rd), "gasrd -c -s 0x%x %zu\r", addr, n);
+
+rd_crc_err:
+	if (crc_rty_cnt > RETRY_NUM)
+		exit(1);
+
+	ret = write(udev->fd, gas_rd, strlen(gas_rd));
```

Comment:
I still want to separate out the common parts of cli_control, uart_memcpy_from_gas and uart_gas_write. Probably into a send_cmd() and a read_resp_line() or something similar. Right now the three functions are all very long, complicated, doing very similar things and are hard to review for correctness. More functions and loops, less goto mess.

Reply:
I have this thoughts before.
But actually, for send_cmd(), it is mainly to invoke syscall write,  just with different input str of cli cmd.

And for read_resp_line(), it is for syscall read.
Besides these two same things, they have different scan template and expected match numbers.
Which will introduce variable length parameters funs, make thing complicated with limited gain.

Reply:
I was actually thinking a variable length parameter function would be a great idea for send_cmd().

read_resp_line() would read until the next prompt, encompassing the waiting for a full line. The calling function can then parse the prompt as it does now.

Not repeating yourself is a huge gain and your existing code is already too complicated..

Reply:
sorry, what i think is:
For send_cmd it is fix lenght parameter function, with fd, str for send.
For response read and check function, it is variable length parameter function, which need different variable for the sscanf output.

For send_cmd, it is just a wrapper for system call write
For read, it is a wrapper for system call read and sscanf.
I still feel it will make simple thing more complex(especially variable length param function).
 
After the command, FW keep send the response, SW determine if the end is comming by check several pattern combination in the entire response string.
Current code is read the entire respone, instead of line by line.

Reply:
read_resp_line() should just read an entire prompt (ie everything up to the '>') the caller can then do subsequent parsing. Alternatively, you could pass in a different function to parse each line. This is more than a wrapper around read and sscanf as it should handle getting partial reads back.

send_cmd() can take vargs so you can do something like:

```ret = send_cmd(dev, "gasrd -c -s 0x%x %zu\r", addr, n);```

It will wrap the sprintf and the write nicely.

Your code is already too complex using a huge **mess** of gotos. I guarantee if you do it right it will be simpler and easier to read and **much** easier to maintain and expand upon.

Reply:
add the read_resp_line function. pls comments.
will add send_cmd function later.